### DISCUSSION TITLE
How do create a custom shader to render graphic fills?

### DISCUSSION THREAD

**Discussion by sketchpunk** [pixijs/pixijs#7728] [Q&A] [answered]
What I'm looking to do is be able to render the fill of Graphics with a Screen Space Procedural Textures. I dont see any examples of how something like this can be done.

**Comment by ivanpopelyshev** [comment, marked as answer]
Usually its done through BatchRenderer so you can batch those graphics.

=====
Way 1

https://www.pixiplayground.com/#/edit/qqx4jdwAQbxlxCbNTf0oL

However im unsure how to do it with uniforms, if you look in sources - it needs its own BatchShaderGenerator , that should have necessary uniforms inside.

====
Way 2

Alternatively, you can assign shader `graphics.shader = ...`  make sure that shader has all necessary attributes:

https://github.com/pixijs/pixijs/blob/7ee2be5cf88bb990166db10cfb017e37aa2cbbb1/packages/graphics/src/Graphics.ts#L988

To make sure it always goes into `renderDirect`, you can do something like `graphics.geometry.isBatchable = () => { return false};`

In that case you can pass uniforms in shader, regularly.

Required attributes list: https://github.com/pixijs/pixijs/blob/7ee2be5cf88bb990166db10cfb017e37aa2cbbb1/packages/core/src/batch/BatchGeometry.ts#L39

Vertex shader: https://github.com/pixijs/pixijs/blob/dev/packages/core/src/batch/texture.vert

Fragment: https://github.com/pixijs/pixijs/blob/dev/packages/core/src/batch/texture.frag , you dont need samplers because, duh, you want a generational texture.

**Reply by ivanpopelyshev** [reply]
Why its not in  examples? - no one actually cared enough to make it. 
Why there's no ability to add shader in Graphics? - it was done several times in different projects but so far there were no PR's for it.

If you figure out how to make better API's and examples for this thing - please share, because without community code this part will be as it is now.

**Reply by sketchpunk** [reply]
Thank you so much for the info. I have some questions about what you shared. The BatchRenderer might be the way to go, but using the playground, I took out %forloop% and everything stopped working. Not sure what that does, its not a GLSL thing. How do I change it so it only applies one color to everything it renders without any textures or those %% syntax things?

So the second solution is to Take the existing shader for Graphics and apply whatever changes I want, is that the gist of it?  Whats renderDirect? Looks like its disabling batching? I would rather not, I would like to keep the batching but ideally change the basic color with a pattern I'd generate instead. This solution might be good too as I would like to keep Graphics to render as it does now with strokes, just change the fill from a solid color to a pattern.

If I get something working well, I'm totally in for making an example that can be added to the Examples page. Is there a process available for me to look through to see how to go about contributing? 

Thanks Again.

**Reply by ivanpopelyshev** [reply]
>  I took out %forloop% and everything stopped working

because shader generator wants it bad. You have to swap shader generator :) While you are at it - you can add any uniforms you want there

**Comment by ivanpopelyshev** [comment]
You can also look in the thread here: https://www.html5gamedevs.com/topic/47664-procedural-texture-shader-batching/

**Reply by sketchpunk** [reply]
thanks. I managed to get a Shader and a BatchRenderer Example working yesterday. I wouldn't mind fixing it up and posting it as an example for pixi. Do I just dump it in playground or do I have to do something else to contribute it to the examples area?

**Reply by ivanpopelyshev** [reply]
Do it in playground, i'll re-save it and publish in examples later :) several people donated their code, i'll cook something up

**Comment by sketchpunk** [comment]
Here's some examples I promised.

// Graphic Shader that creates a Screenspace Fill Pattern while leaving stroke's solid
https://www.pixiplayground.com/#/edit/tYcT_u_Lc4mPPQbfUPASN

// Same Pattern Fill Shader, but this time running in a Custom Batch Renderer & Shader Generator
https://www.pixiplayground.com/#/edit/1lREoA-hv8TqteYDil5jE

// Bonus: A Nice Grid Background Shader
https://www.pixiplayground.com/#/edit/guq3xCkrC_NFhbT81xc-Q

**Reply by ivanpopelyshev** [reply]
Awesome! OK, i'm posting it to our slack to guys who are making pixijs guides ( https://github.com/pixijs/guides ) 
Btw, do you want invite to slack? I just need an email for it

**Reply by miltoncandelero** [reply]
What does shader do on graphics? I only used shaders on either filters or mesh 🤔🤔

**Reply by sketchpunk** [reply]
sure, {myGitName}-atGmail  Sorry for being cyptic, trying to avoid bots.

**Reply by sketchpunk** [reply]
The shader I made to run on Graphics is to add a procedurally generated pattern instead of a solid color. This way I can avoid using textures.

**Reply by matjazPrijatelj** [reply]
Can you please fix the first example (direct graphic.shader example) and disable editing since links are publicly allowed to edit it.
